ext/test/key_provider/key_provider.c

```diff
@@ -0,0 +1,403 @@
+      kp->state.current_lsn, CLOCK_SECS(kp->state.key_time), kp->state.key_size);
+
+    LOG_INFO(
+      kp, session, "Loading key for LSN=%" PRIu64 ", size=%" PRIzu, crypt->r.lsn, crypt->keys.size);
+
+    kp_set_key(kp, crypt);
+
+    return (0);
+}
+
+/*
+ * kp_key_expired --
+ *     Check if the current key has expired based on the configured expiration time.
+ */
+static bool
+kp_key_expired(KEY_PROVIDER *kp)
+{
+    if (kp->key_expires == 0)
+        return (false); /* Key does not expire */
+
+    const clock_t now = clock();
+    double elapsed_sec = CLOCK_SECS(now - kp->state.key_time);
+
+    return (elapsed_sec >= kp->key_expires);
+}
+
+/*
+ * kp_rotate_key --
+ *     Rotate the current key by generating a new key with a repeating alphabet pattern.
+ */
+static int
+kp_rotate_key(KEY_PROVIDER *kp)
+{
+    /* Calculate new key size with 20% random fluctuation */
+    const size_t base_size = 1024;
+    const int fluctuation = (rand() % 41) - 20; /* -20% to +20% */
+    const size_t new_key_size = base_size + (size_t)((int)base_size * fluctuation / 100);
+
+    /* Allocate new key buffer */
+    uint8_t *new_key = malloc(new_key_size);
+    if (new_key == NULL)
+        return (ENOMEM);
+
+    /* Fill with repeating alphabet pattern */
+    const char *alphabet = "abcdefghijklmnopqrstuvwxyz";
+    const size_t alphabet_len = strlen(alphabet);
+
+    /* Fill buffer by repeatedly copying the alphabet pattern */
+    size_t remaining = new_key_size;
+    size_t offset = 0;
+    while (remaining > 0) {
+        const size_t copy_len = (remaining >= alphabet_len) ? alphabet_len : remaining;
+        memcpy(new_key + offset, alphabet, copy_len);
+        offset += copy_len;
+        remaining -= copy_len;
+    }
+
+    /* Free old key and update state */
+    kp_free_key(kp);
+    kp->state.current_key = new_key;
+    kp->state.key_size = new_key_size;
+    kp->state.key_time = clock();
+    kp->state.current_lsn = 0; /* New key, no LSN yet */
+
+    return (0);
+}
+
+/*
+ * kp_get_key --
+ *     Fetch the latest key for checkpoint writes. The WT_CRYPT_KEYS::keys::size should be set to
+ *     zero if the key has not changed. This is called by WiredTiger when creating a checkpoint to
+ *     get the current encryption key.
+ */
+static int
+kp_get_key(WT_KEY_PROVIDER *wtkp, WT_SESSION *session, WT_CRYPT_KEYS *crypt)
+{
+    KEY_PROVIDER *kp = (KEY_PROVIDER *)wtkp;
+    LOG_DEBUG(kp, session, "Current key: LSN=%" PRIu64 ", key_time=%.2f, size=%" PRIzu,
+      kp->state.current_lsn, CLOCK_SECS(kp->state.key_time), kp->state.key_size);
+
+    /*
+     * Real key provider may rotate the key independently of the get_key calls. In the mock
+     * implementation the key is rotated only when its size has been requested. This is to prevent
+     * key rotation between paired get_key calls: first requesting the size, then filling the data.
+     */
+    if (crypt->keys.data == NULL && kp_key_expired(kp)) {
```

Comment:
Code using: `crypt->keys.data` makes me uncertain about which side (WT or the key_provider module) "owns" and sets crypt->keys.data.  I would have thought that crypt->keys.data is allocated by key_provider (because it knows the size), and is freed by the key_provider on the call to on_key_update.  But this usage does not seem to follow that.